b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+
+import sys
+
+from abc import ABCMeta, abstractclassmethod, abstractmethod
+from collections import Counter
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from queue import Queue
+from typing import ClassVar, Dict, Iterator, Optional, Tuple, Type
+
+from ..api import B2Api
+from ..bucket import Bucket
+from ..encryption.setting import EncryptionMode
+from ..file_lock import NO_RETENTION_FILE_SETTING, LegalHold
+from ..file_version import FileVersion
+from ..scan.folder import B2Folder
+from ..scan.path import B2Path
+from ..scan.policies import DEFAULT_SCAN_MANAGER, ScanPoliciesManager
+from ..scan.report import Report
+from ..scan.scan import zip_folders
+from .setting import ReplicationRule
+from .types import ReplicationStatus
+
+
+@dataclass(frozen=True)
+class AbstractReplicationAttrs(metaclass=ABCMeta):
+    """
+    Some attributes of files which are meaningful
+    for replication monitoring and troubleshooting.
+    """
+
+    @abstractclassmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'AbstractReplicationAttrs':
+        pass
+
+
+@dataclass(frozen=True)
+class ReplicationAttrs(AbstractReplicationAttrs):
+    """
+    Some attributes of source and destination files and their relations
+    which are meaningful for replication monitoring and troubleshooting.
+    """
+
+    # source attrs
+    source_replication_status: Optional[ReplicationStatus] = None
+    source_has_hide_marker: Optional[bool] = None
+    source_has_sse_c_enabled: Optional[bool] = None
+    source_has_large_metadata: Optional[bool] = None
+    source_has_file_retention: Optional[bool] = None
+    source_has_legal_hold: Optional[bool] = None
+
+    # destination attrs
+    destination_replication_status: Optional[ReplicationStatus] = None
+
+    # source & destination relation attrs
+    metadata_differs: Optional[bool] = None
+
+    LARGE_METADATA_SIZE: ClassVar[int] = 2048
+
+    @classmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'ReplicationAttrs':
+        params = {}
+
+        if source_file:
+            source_file_version = source_file.selected_version
+            params.update(
+                {
+                    'source_replication_status':
+                        source_file_version.replication_status,
+                    'source_has_hide_marker':
+                        source_file.is_visible(),
+                    'source_has_sse_c_enabled':
+                        source_file_version.server_side_encryption.mode == EncryptionMode.SSE_C,
+                    'source_has_large_metadata':
+                        source_file_version.headers_size >= cls.LARGE_METADATA_SIZE,
```

Comment:
I thinking whether the logic which determines whether a `FileVersion` object has large metadata size should be implemented in:
a) replication.monitoring.ReplicationAttrs
b) FileVersion
and I think it might need to be b). Then `_headers_size` can become private.

It's kind of a special case, all of the others are fine like they are now.